### DISCUSSION TITLE
Why does (= 0 nil) return false, but (= nil 0) return true?

### DISCUSSION THREAD

**Discussion by oadbd** [janet-lang/janet#1295] [Q&A] [answered]
Hi, new to janet.

In the docs under comparision operators it says 
`(= nil nil)      # true -- nil always equal to itself and only itself
`

However, I find if I do something like 
`(= (string/find "foo" "bar") 0)` this will return true

Switching it up to the following works as expected.
`(= 0 (string/find "foo" "bar"))`

**Comment by bakpakin** [comment]
What platform are you on? Cannot reproduce on my machine.

```janet
Janet 1.31.0-local windows/x64/msvc - '(doc)' for help
repl:1:> (= 0 (string/find "foo" "bar"))
false
repl:2:> (= 0 (string/find "foo" "bar") 0)
false
repl:3:> (= (string/find "foo" "bar") 0)
false
repl:4:>
```

**Reply by oadbd** [reply]
I'm on a macbook pro m1, with janet 1.30-meson installed through brew

**Reply by oadbd** [reply]
<img width="519" alt="image" src="https://github.com/janet-lang/janet/assets/188016/cb4e384c-7808-48e3-941c-8dd00fd46dd7">

**Reply by bidulock** [reply]
I did just try compiling the latest version also with the same results.
```janet
Janet 1.31.0-910b9cf1 macos/aarch64/clang - '(doc)' for help
repl:1:> (= nil 0)
true
repl:2:>
```

Just trying out a few other releases a bit randomly 1.15.0 has the same result, but 1.0.0 seems to work

```janet
Janet 1.0.0-d66f8333  Copyright (C) 2017-2019 Calvin Rose
janet:0:> (= nil 0)
false
janet:10:>
```

Since the expected behaviour is `(= nil 0)` to be false, I can spend some time tomorrow bisecting this to find the commit that introduced it and create an issue in github if that helps.

**Reply by ianthehenry** [reply]
It looks like `eqim`/`neqim` don't check the type of the LHS, and on platforms without nanboxing casting `nil` to a double gives you numeric zero -- `(= false 0)` is also true for the same reason. If you don't use an immediate 0 it works as expected -- e.g. `(= nil (- 1 1))` is false. With nanboxing I think this is always safe because there's no way to write an immediate NaN.

(I think if you could figure out how to write the floating point number represented by `0x1` as a Janet literal you could get `(= true ...)` to give you the wrong result too... which would be kinda fun but not worth figuring out how to print the full decimal expansion...)

**Reply by sogaiu** [reply, marked as answer]
May be these are the lines in question (just following up on ianthehenry's analysis)?

https://github.com/janet-lang/janet/blob/910b9cf1fdd202aa780a75285f84c87bbd3ca659/src/core/vm.c#L863-L865

https://github.com/janet-lang/janet/blob/910b9cf1fdd202aa780a75285f84c87bbd3ca659/src/core/vm.c#L871-L873

**Reply by oadbd** [reply]
I did run a bisect and I believe the issue shows up in this commit https://github.com/janet-lang/janet/commit/aefde67aa298a6897607b8d97a83bc6f0889a7d8 

Making this change does fix the issue, but I doubt this is the real fix, as it looks like @ianthehenry and @sogaiu have already narrowed things down. But I'll leave this here if it helps. 

```diff
index 46ecfeb0..4a211293 100644
--- a/src/core/cfuns.c
+++ b/src/core/cfuns.c
@@ -349,7 +349,7 @@ static JanetSlot do_lte(JanetFopts opts, JanetSlot *args) {
     return compreduce(opts, args, JOP_LESS_THAN_EQUAL, 0, 0);
 }
 static JanetSlot do_eq(JanetFopts opts, JanetSlot *args) {
-    return compreduce(opts, args, JOP_EQUALS, JOP_EQUALS_IMMEDIATE, 0);
+    return compreduce(opts, args, JOP_EQUALS, 0, 0);
 }
 static JanetSlot do_neq(JanetFopts opts, JanetSlot *args) {
     return compreduce(opts, args, JOP_NOT_EQUALS, JOP_NOT_EQUALS_IMMEDIATE, 1);
```

**Comment by bakpakin** [comment]
Should be fixed in 7b4c3bdbcc2d23bd2fb5349026d0b8531f67a694
